File: 백준/Gold/3079. 입국심사/입국심사.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static ArrayList<Integer> arr = new ArrayList<>();
	public static void main (String[] args) throws java.lang.Exception {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringTokenizer st;
	    
	    st = new StringTokenizer(br.readLine());
	    int n = Integer.parseInt(st.nextToken());
	    int m = Integer.parseInt(st.nextToken());
	    
	    long low = 1, high = 1000000000;
	    
	    for(int i = 0; i < n; i++){
	        int tmp = Integer.parseInt(br.readLine());
	        arr.add(tmp);
	        high = Math.min(high, (long)tmp);
	    }
	    
	    Collections.sort(arr, (a, b) -> Integer.compare(a, b));
	    high *= m;
	    
	    long res = high;
	    while(low <= high){
	        long mid = (low + high) / 2;
	        
	        long cnt = 0;
	        for(int i = 0; i < arr.size(); i++){
	            cnt += mid / arr.get(i);
	            if(cnt >= m)
	                break;
	        }
	        
	        if(cnt >= m){
	            res = mid;
	            high = mid - 1;
	        }
	        else{
	            low = mid + 1;
	        }
	    }
	    
	    System.out.println(res);
	}
}

File: 백준/Gold/3079. 입국심사/입국심사.cc
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    long long low = 1, high = 1000000000;
    cin >> n >> m;
    
    vector<int> pass;
    
    int tmp;
    for(int i = 0; i < n; i++){
        cin >> tmp;
        pass.push_back(tmp);
        high = min(high, (long long)tmp);
    }
    
    sort(pass.begin(), pass.end());
    
    long long res = 0;
    high *= m;
    while(low <= high){
        long long mid = (low + high) / 2;
        
        long long cnt = 0;
        
        for(int i = 0; i < pass.size(); i++){
            cnt += mid / pass[i];
            if(cnt >= m)
                break;
        }
        
        if(cnt >= m){
            res = mid;
            high = mid - 1;
        }
        else{
            low = mid + 1;
        }
    }
    cout << res;
    return 0;
}